cou/apps/base.py

```diff
@@ -621,22 +637,41 @@ def _need_current_channel_refresh(self, target: OpenStackRelease) -> bool:
         :return: True if needs to refresh, False otherwise
         :rtype: bool
         """
-        return bool(self.can_upgrade_to) and self.channel_codename <= target
+        return bool(self.can_upgrade_to) and self.current_channel_os_release <= target
 
     def _get_upgrade_charm_step(self, target: OpenStackRelease) -> UpgradeStep:
         """Get step for upgrading the charm.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
+        :raises ApplicationError: When the current channel is ahead from expected and the target.
         :return: Step for upgrading the charm.
         :rtype: UpgradeStep
         """
-        if self.channel != self.target_channel(target):
+        channel = self.expected_current_channel(target) if self.need_crossgrade else self.channel
+
+        if (
+            self.current_channel_os_release <= self.current_os_release
```

Comment:
I changed to `channel_o7k_release` and `o7k_release`. I don't think `payload_o7k_release` is a good name because it's not always that the OpenStack release is decided by the payload.

Obs: changing from `os` to `o7k` was a lot of lines

Reply:
yeah... it would have probably been better to do the renaming in a separate PR